File: src/py/2416. Sum of Prefix Scores of Strings/2416. Sum of Prefix Scores of Strings.py
class Trie:
    def __init__(self):
        self.children = [None] * 26
        self.counts = 0

class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        tt = Trie()
        base = ord('a')
        
        for word in words:
            node = tt
            for c in word:
                idx = ord(c) - base
                if node.children[idx] is None:
                    node.children[idx] = Trie()
                node = node.children[idx]
                node.counts += 1
        
        out = []
        for word in words:
            node = tt
            score = 0
            for c in word:
                idx = ord(c) - base
                if node.children[idx] is not None:
                    node = node.children[idx]
                    score += node.counts
            out.append(score)
            
        return out

File: src/cpp/2416. Sum of Prefix Scores of Strings/2416. Sum of Prefix Scores of Strings.cpp
class Trie {
public:
    vector<Trie*> children;
    int counts;
    Trie(): children(26, nullptr), counts{0} {}
};


class Solution {
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        auto root {new Trie()};
            
        for (const auto& word : words) {
            auto iter {root};
            for (const auto& c : word) {
                auto idx {c - 'a'};
                if (iter->children[idx] == nullptr) {
                    iter->children[idx] = new Trie();
                }
                iter = iter->children[idx];
                iter->counts += 1;
            }
        }
        
        
        vector<int> out;
        for (auto word : words) {
            auto iter {root};
            auto score {0};
            for (const auto& c : word) {
                auto idx {c - 'a'};
                iter = iter->children[idx];
                score += iter->counts;
            }
            out.push_back(score);
        }
        return out;
    }
};
